[FLINK-4022] [kafka] Partition and topic discovery for FlinkKafkaConsumer

tzulitai opened the pull request:
This PR adds the required internals to allow partition and topic regex pattern discovery in the `FlinkKafkaConsumer`.

It doesn't expose a new constructor that accepts regex topic patterns yet. I propose to expose that with https://issues.apache.org/jira/browse/FLINK-5704 (deprecate the original FlinkKafkaConsumer constructors in favor of new ones with new offset behaviours). For this reason, I also propose to update the Kafka documentation when the new constructors are added.

## Design

Some description to ease review:

- `AbstractPartitionDiscoverer`:
An `AbstractPartitionDiscoverer` is a stateful utility instance that remembers what partitions are discovered already. It also wraps the logic for partition-to-subtask assignment. The main `run()` method now has a discovery loop that calls `AbstractPartitionDiscoverer#discoverPartitions()` on a fixed interval. This method returns only new partitions that should be subscribed by the subtask.
The returned partitions are used to invoke `AbstractFetcher#addDiscoveredPartitions(...)` on the fetcher.
On a fresh startup, `AbstractPartitionDiscoverer#discoverPartitions()` is also used to fetch the initial seed startup partitions in `open()`.

- `AbstractFetcher#addDiscoveredPartitions(...)`
The fetcher now has an `unassignedPartitionsQueue` that contains discovered partitions not yet assigned to Kafka consumers to be read. Whenever `addDiscoveredPartitions(...)` is called on the fetcher, the fetcher will create the state holders for the partitions, and add the partitions to the queue.
Concrete implementations of the fetcher should continuously poll this queue in the fetch loop. If partitions are found from the queue, they should be assigned for consuming.

- Concrete fetchers continuously polls the queue in `runFetchLoop()`
For 0.8, this simply means that the original `unassignedPartitionsQueue` in Kafka08Fetcher is moved to the base abstract fetcher class. Nothing else is touched.
For 0.9+, queue polling and partition reassignment for the high-level consumer happens in `KafkaConsumerThread`.

With this design, partition discovery is split into 2 separate concerns:
- Fetchers simply polls the queue for new partitions to read from in their fetch loop.
- The main run loop serves as the discovery loop. Whenever new partitions are found, they are added to the queue. Initial partitions that are found on fresh startups (retrieved before entering the discovery loop) are treated as seeded elements in the queue.

## Limitations

For the partition discovery to work properly after restores, the previous state had to be migrated to use union list states instead (`OperatorStateStore#getUnionListState()`).

One limitation that comes with this migration is that if the current run was restored from old state (Flink 1.1 / 1.2), then partition discovery is disabled. In order to use partition discovery, after the job starts the user must do a manual snapshot (with the new union state) & restore.

tzulitai commented:
@haohui FYI - tagging you as we discussed about this feature offline

aljoscha commented on flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractPartitionDiscoverer.java:
```diff
@@ -0,0 +1,254 @@
+			try {
+				List<KafkaTopicPartition> newDiscoveredPartitions;
+
+				// (1) get all possible partitions, based on whether we are subscribed to fixed topics or a topic patern
```
Typo "patern"

aljoscha commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThread.java:
```diff
@@ -219,6 +209,15 @@ public void run() {
+						reassignPartitions(newPartitions);
+					}
+				} catch (AbortedReassignmentException e) {
+					continue;
```
Why is it necessary to continue here, instead of continuing with the rest of the loop?

tzulitai replied:
the reason is that the `AbortedReassignmentException` is thrown only if the Kafka consumer was `wakeup` before the reassignment started and that wakeup effect was consumed by an operation during the reassignment.

Now, if the loop just continues, the next Kafka operation (the `poll`) would not be woken up because the wakeup was already consumed.
So, effectively, it would mean the same thing to do this in the catch block instead of falling through the loop:
```
} catch (AbortedReassignmentException e) {
consumer.wakeup();
}
```

I'm just eagerly continuing instead of restoring the wakeup. Does that make sense?

tzulitai replied:
One extra remark:

the external `wakeup` call was intended to un-block the main thread from any Kafka blocking operation, and commit offsets as soon as possible (which happens at the start of the loop). It's also for making sure that the main thread stops work as soon as possible so that there isn't any strange long job shutdown issues (which occurred before).

Hence it's important to restore effective wakeup calls (or mimic their behaviour) and fall through the loop when an external operation did call wakeup.

aljoscha replied:
Thanks! that makes sense

aljoscha commented on flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThreadTest.java:
```diff
@@ -0,0 +1,842 @@
+	/**
+	 * A testable {@link KafkaConsumerThread} that injects multiple latches exactly before and after
+	 * partition reassignment, so that tests are eligible to setup various conditions before the reassignment happens
+	 * and inspect reqssignment results after it is completed.
```
Typo "reqssignment"

aljoscha commented on flink-connectors/flink-connector-kafka-0.9/src/test/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThreadTest.java:
```diff
@@ -0,0 +1,842 @@
+
+		testThread.setOffsetsToCommit(new HashMap<TopicPartition, OffsetAndMetadata>());
+
+		// the wakeup the setOffsetsToCommit() call should have been buffered, and not called on the consumer
```
should be "the wake in the setOffset..."?

tzulitai commented on flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java:
```diff
@@ -424,65 +485,136 @@ public void run(SourceContext<T> sourceContext) throws Exception {
+				public void run() {
+					try {
+						// run the fetcher' main work method
+						kafkaFetcher.runFetchLoop();
```
Hmm, there actually isn't any good reason that this is required, as I can think of.

one point regarding non-main thread emitting stuff: the Kafka 0.8 fetcher actually had always been emitting elements from different threads. So I didn't really assume which thread (main or separate) runs the fetcher loop and which one runs the discovery loop.

but I think it's also ok to swap this here

tzulitai commented:
@aljoscha thanks for the review! Yes, the Kafka 0.8 code hasn't been touched much, and the `KafkaPartitionDiscoverer08` is mainly just a copy of the previous partition fetching code with only minor refactoring.

haohui commented:
Any updates on this? Would love to try this out :-)

tzulitai commented:
@haohui you should be able to the PR now, would be great to hear feedback on this.
As of now I'm not quite sure whether this will make it into 1.3 though; I think we'll have to see about the status of the 1.3 release in a while.

tzulitai commented on flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java:
```diff
@@ -503,23 +644,30 @@ public void close() throws Exception {
+
+			// migrate from 1.2 state, if there is any
+			for (Tuple2<KafkaTopicPartition, Long> kafkaOffset : oldRoundRobinListState.get()) {
+				restoredFromOldState = true;
```
You're right, this is definitely problematic. I think we'll have to block this feature until we have the required state information accessible before we can merge this then.

tzulitai commented:
Hi @haohui, unfortunately we might not be able to merge this for 1.3. The reason is that the feature requires a migration from partitionable list state to union list state, which currently can not be properly done, as can be seen in Aljoscha's inline comment.

haohui commented:
@tzulitai we are interested in putting this feature in production soon. Do you think whether it is possible for us to get a workaround for 1.3? Thanks.

tzulitai commented:
@haohui I can not guarantee this, but I can try revisiting this for a workaround after I've done some testing on the current RC1.

tzulitai commented:
Hi @haohui, I revisited this, but I think there is no easy workaround for this, unfortunately.
This can only be properly merged in 1.4 after some additional requirements are added to Flink.

However, if you know that during Flink 1.2 your FlinkKafkaConsumer was not rescaled before, then in that case specifically you should be able to use this patch. It would however require a custom build of the connector on your side.

tzulitai commented:
Hi @aljoscha, I've rebased this PR, fixed the previous blocking issue (discovery enabling & restoring from round-robin list state), and added documentation for the limitation. Could you have another look? Thanks!

aljoscha commented on flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java:
```diff
@@ -189,6 +192,8 @@
 	 * @param topicPattern the topic pattern to subscribe to (null, if using fixed topics)
 	 * @param deserializer The deserializer to turn raw byte messages into Java/Scala objects.
+	 * @param discoveryIntervalMillis the topic / partition discovery interval, in
+	 *                                milliseconds (0 if discovery is disabled).
```
Here it says `0` for disabled while in other places it's `Long.MIN_VALUE`.

tzulitai commented:
Thanks for the review @aljoscha. Will address the comment and merge after Travis gives green :)

tzulitai commented:
Travis passed, merging ...

tzulitai commented:
(please ignore the batch of commits. I'm just using this branch for an overall Travis run before committing the batch; personal accounts get time-outs too constantly)
